packages/core/content-releases/server/src/services/release.ts

```diff
@@ -460,166 +460,192 @@ const createReleaseService = ({ strapi }: { strapi: LoadedStrapi }) => ({
-
-    // Only publish the release if all action updates are applied successfully to their entry, otherwise leave everything as is
-    await strapi.db.transaction(async () => {
-      // First we publish all the singleTypes
-      for (const { uid, action, id } of singleTypeActions) {
-        // @ts-expect-error - populateBuilderService should be a function but is returning service
-        const populate = await populateBuilderService(uid).populateDeep(Infinity).build();
+      if (releaseWithPopulatedActionEntries.releasedAt) {
+        throw new errors.ValidationError('Release already published');
+      }
 
-        const entry = await strapi.entityService.findOne(uid, id, { populate });
+      if (releaseWithPopulatedActionEntries.actions.length === 0) {
+        throw new errors.ValidationError('No entries to publish');
+      }
 
-        try {
-          if (action === 'publish') {
-            await entityManagerService.publish(entry, uid);
-          } else {
-            await entityManagerService.unpublish(entry, uid);
+      /**
+       * We separate publish and unpublish actions, grouping them by contentType and extracting only their IDs. Then we can fetch more data for each entry
+       * We need to separate collectionTypes from singleTypes because findMany work as findOne for singleTypes and publishMany can't be used for singleTypes
+       */
+      const collectionTypeActions: {
+        [key: UID.ContentType]: {
+          entriestoPublishIds: ReleaseAction['entry']['id'][];
+          entriesToUnpublishIds: ReleaseAction['entry']['id'][];
+        };
+      } = {};
+      const singleTypeActions: {
+        uid: UID.ContentType;
+        id: ReleaseAction['entry']['id'];
+        action: ReleaseAction['type'];
+      }[] = [];
+      for (const action of releaseWithPopulatedActionEntries.actions) {
+        const contentTypeUid = action.contentType;
+
+        if (strapi.contentTypes[contentTypeUid].kind === 'collectionType') {
+          if (!collectionTypeActions[contentTypeUid]) {
+            collectionTypeActions[contentTypeUid] = {
+              entriestoPublishIds: [],
+              entriesToUnpublishIds: [],
+            };
           }
-        } catch (error) {
-          if (
-            error instanceof errors.ApplicationError &&
-            (error.message === 'already.published' || error.message === 'already.draft')
-          ) {
-            // We don't want throw an error if the entry is already published or draft
+
+          if (action.type === 'publish') {
+            collectionTypeActions[contentTypeUid].entriestoPublishIds.push(action.entry.id);
           } else {
-            throw error;
+            collectionTypeActions[contentTypeUid].entriesToUnpublishIds.push(action.entry.id);
           }
+        } else {
+          singleTypeActions.push({
+            uid: contentTypeUid,
+            action: action.type,
+            id: action.entry.id,
+          });
         }
       }
```

Comment:
Not for this pr but idea for refactor:

What do you think about separating this into two functions

`getCollectionTypePublishActions` returns:
```
[key: UID.ContentType]: {
    entriestoPublishIds: ReleaseAction['entry']['id'][];
    entriesToUnpublishIds: ReleaseAction['entry']['id'][];
  }
```

`getSingleTypePublishActions` returns:

```
{
  uid: UID.ContentType;
  id: ReleaseAction['entry']['id'];
  action: ReleaseAction['type'];
}[]
```

Reply:
I like it, good candidate to refactor in the v5 migration of releases maybe?

I will add a note for myself